grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcServiceBuilder.java

```diff
@@ -324,7 +324,90 @@ public GrpcServiceBuilder addService(String path, BindableService bindableServic
             return addService(path, interceptor, methodDescriptor);
         }
         registryBuilder.addService(path, bindableService.bindService(), methodDescriptor,
-                                   bindableService.getClass());
+                                   bindableService.getClass(), null);
+        return this;
+    }
+
+    /**
+     * Decorates a gRPC {@link BindableService} with the given decorators, in the order of iteration.
+     * For more details on decorator behavior, please refer to the following document.
+     *
+     * @see <a href="https://armeria.dev/docs/server-grpc#decorating-a-grpcservice">Decorating a GrpcService</a>
+     */
+    @UnstableApi
+    public GrpcServiceBuilder addService(
+            BindableService bindableService,
+            Iterable<? extends Function<? super HttpService, ? extends HttpService>> decorators) {
+        requireNonNull(bindableService, "bindableService");
+        if (bindableService instanceof ProtoReflectionService) {
+            throw new IllegalArgumentException("ProtoReflectionService should not be used with decorators.");
+        }
+        requireNonNull(decorators, "decorators");
+        registryBuilder.addService(bindableService.bindService(), bindableService.getClass(), decorators);
+        return this;
+    }
```

Comment:
How about making other variants delegate to this method? We can allow `ProtoReflectionService` if `decorators` is empty.
For example:
```suggestion
    public GrpcServiceBuilder addService(
            BindableService bindableService,
            Iterable<? extends Function<? super HttpService, ? extends HttpService>> decorators) {
        requireNonNull(bindableService, "bindableService");
        requireNonNull(decorators, "decorators");
        if (bindableService instanceof ProtoReflectionService)) {
            if (!Iterables.isEmpty(decorators) {
               throw new IllegalArgumentException("ProtoReflectionService should not be used with decorators.");
            }
            return addService(ServerInterceptors.intercept(bindableService,
                                                           newProtoReflectionServiceInterceptor()));
        }
        registryBuilder.addService(bindableService.bindService(), bindableService.getClass(), decorators);
        return this;
    }
```